tools/python_api/src_cpp/py_udf.cpp

```diff
@@ -0,0 +1,178 @@
+                }
+            }
+            auto numSelectedValues = 0u;
+            auto selectedPositionsBuffer = selVector.getMultableBuffer();
+            for (auto i = 0u; i < params[unFlatVectorIdx]->state->selVector->selectedSize; ++i) {
+                auto resultPos = params[unFlatVectorIdx]->state->selVector->selectedPositions[i];
+                py::list pyParams;
+                for (const auto& param : params) {
+                    auto paramPos = param->state->isFlat() ?
+                                        param->state->selVector->selectedPositions[0] :
+                                        resultPos;
+                    auto value = param->getAsValue(paramPos);
+                    auto pyValue = PyQueryResult::convertValueToPyObject(*value);
+                    pyParams.append(pyValue);
+                }
+                auto pyResult = udf(*pyParams);
+                auto resultValue =
+                    PyConnection::transformPythonValueAs(pyResult, LogicalType::BOOL().get());
+                selectedPositionsBuffer[numSelectedValues] = resultPos;
+                numSelectedValues += resultValue.getValue<bool>();
+            }
+            selVector.selectedSize = numSelectedValues;
+            return numSelectedValues > 0;
+        };
+}
+
+function_set PyUDF::toFunctionSet(const std::string& name, const py::function& udf,
+    const py::list& params, const std::string& retType) {
+    auto pySignature = analyzeSignature(udf);
+    auto explicitParams = pyListToParams(params);
+    if (explicitParams.size() > 0) {
+        if (explicitParams.size() != pySignature.params.size()) {
+            throw RuntimeException(
+                "Explicit parameters and Function parameters must be the same length");
+        }
+        pySignature.params = explicitParams;
+    }
+    for (auto i : pySignature.params) {
+        if (i == LogicalTypeID::ANY) {
+            throw RuntimeException(
+                "Parameters must be annotated or explicitly given, and cannot be ANY");
+        }
+    }
+    if (retType != "") {
+        auto explicitRetType = LogicalTypeUtils::fromStringToID(retType);
+        pySignature.retval = explicitRetType;
+    }
+    if (pySignature.retval == LogicalTypeID::ANY) {
+        throw RuntimeException(
+            "Return value must be annotated or explicitly given, and cannot be ANY");
+    }
+
+    // TODO maxwell: support nested parameters
+    for (auto i : pySignature.params) {
+        if (LogicalTypeUtils::isNested(i)) {
+            throw RuntimeException("Nested type UDFs not implemented yet");
+        }
+    }
+    if (LogicalTypeUtils::isNested(pySignature.retval)) {
+        throw RuntimeException("Nested type UDFs not implemented yet");
+    }
+
+    function_set definitions;
+    definitions.push_back(std::make_unique<ScalarFunction>(name, pySignature.params,
+        pySignature.retval, getUDFExecFunc(udf), getUDFSelectFunc(udf), nullptr /* bindFunc */));
```

Comment:
Some functions don't have select function, (e.g. arithmetic operation +). I think you can skip udfselectfunc() for now.